objc/src/main/java/org/robovm/objc/ObjCObject.java

```diff
@@ -318,11 +321,61 @@ public ObjCObjectRef(ObjCObject referent) {
         }
     }
     
+    static class ObjectOwnershipHelper {
+        private static final Map<Long, Object> CUSTOM_OBJECTS = new HashMap<Long, Object>();
+        
+        private static final long retainCount = Selector.register("retainCount").getHandle();
+        private static final long retain = Selector.register("retain").getHandle();
+        private static final long newRetain = Selector.register("original_retain").getHandle();
+        private static final long release = Selector.register("release").getHandle();
+        private static final long newRelease = Selector.register("original_release").getHandle();
+        
+        public static void registerClass(ObjCClass objCClass) {
+            registerCallbackMethod(objCClass.getHandle(), retain, newRetain, "retain", ObjCObject.class, Long.TYPE);
+            registerCallbackMethod(objCClass.getHandle(), release, newRelease, "release", Long.TYPE, Long.TYPE);
+        }
+        
+        private static void registerCallbackMethod (long cls, long selector, long newSelector, String methodName, Class<?>... parameterTypes) {
+            Method method = null;
+            try {
+                method = ObjectOwnershipHelper.class.getDeclaredMethod(methodName, parameterTypes);
+            } catch (Throwable t) {
+                throw new Error(t);
+            }
+            long superMethod = ObjCRuntime.class_getInstanceMethod(cls, selector);
+            long typeEncoding = ObjCRuntime.method_getTypeEncoding(superMethod);
+
+            if (!ObjCRuntime.class_addMethod(cls, selector, VM.getCallbackMethodImpl(method), typeEncoding)) {
+                throw new Error("Failed to init the ObjectOwnershipHelper: class_addMethod(...) failed");
+            }
+            ObjCRuntime.class_replaceMethod(cls, newSelector, ObjCRuntime.method_getImplementation(superMethod), typeEncoding);
+        }
+
+        @Callback
+        private static @Pointer long retain (ObjCObject self, @Pointer long sel) {
+            if (ObjCRuntime.int_objc_msgSend(self.getHandle(), retainCount) == 1) {
```

Comment:
From my tests and observations it seems every object starts with a retain count of 1. I've never seen retain being called with a retain count of 0. Though to be on the save side, we should probably check if it's <= 1.